### DISCUSSION TITLE
rocky-demo.exe doesn't want to start

### DISCUSSION THREAD

**Discussion by kahlenberg** [pelicanmapping/rocky#26] [Q&A / Support] [answered]
Hi,
I succesfully built the `rocky-demo.exe` and set the environment variables `VSG_FILE_PATH`, `ROCKY_FILE_PATH`
 and `PROJ_DATA`, however I receive following error message:

![image](https://github.com/pelicanmapping/rocky/assets/6330648/bbd3e50d-8825-41c7-9461-a5bcb40635b7)

![image](https://github.com/pelicanmapping/rocky/assets/6330648/bda1ab3a-f253-4c05-94cc-e04697f2813e)

All necessary dll files are in the build directory.

**Comment by Harsh-patel00** [comment]
The shaders are present in rocky's_install_directory/shaders folder, make sure you have that folder present. If not, try re-installing rocky.

**Reply by kahlenberg** [reply]
Shaders are in correct directory and `ROCKY_FILE_PATH` points to correct path.
Not only rocky-demo.exe, but rdemo.exe and rsimple.exe are also not working, interstingly rengine.exe works and shows the earth.
I am debugging rdemo.exe and I receive an exception in **C:\workspace\vcpkg\buildtrees\vsg\src\33d994159e-2523849642.clean\src\vsg\vk\Device.cpp**

![image](https://github.com/pelicanmapping/rocky/assets/6330648/bff0c21c-2e9f-49ac-b813-094f83e52f23)

**Reply by gwaldron** [reply]
Really not sure. You may need to step into the code with a debugger, for starters look around InstanceVSG.cpp:253 and make sure it's pulling the environment variables correctly.

**Reply by kahlenberg** [reply]
At InstanceVSG.cpp:253 path is found, I think screenshot in the first post is misleading. After setting the correct paths, I don't receive these warnings about ROCKY_FILE_PATH anymore.
In Debug-session, the problem (exception) occurs at C:\workspace\rocky\src\rocky\vsg\DisplayManager.cpp:74
`    auto commandgraph = vsg::CommandGraph::create(window);`

**Reply by Harsh-patel00** [reply]
Are your layers getting loaded? I was facing errors during Device Creation when layers were not present.

**Reply by kahlenberg** [reply]
How can I check if layers are loaded correctly?

**Reply by Harsh-patel00** [reply]
Rocky will give you warnings in the console regarding layers.

**Reply by gwaldron** [reply]
This code in `rdemo.cpp` checks for a layer error:
```c++
        if (layer->status().failed())
            return layerError(layer);
```

**Reply by gwaldron** [reply]
Since `abort()` was called, you probably ran into a ROCKY_HARD_ASSERT failure. Run from the debugger and it should breakpoint where the assertion occurs. or click "Retry" to see the call stack.

**Reply by kahlenberg** [reply]
> This code in `rdemo.cpp` checks for a layer error:
> 
> ```c++
>         if (layer->status().failed())
>             return layerError(layer);
> ```
It loads layers correctly, it doesn't return at that point.

**Reply by kahlenberg** [reply]
> Since `abort()` was called, you probably ran into a ROCKY_HARD_ASSERT failure. Run from the debugger and it should breakpoint where the assertion occurs. or click "Retry" to see the call stack.

Debugger points the line     `for (auto queueInfo : queueCreateInfos)` in DisplayManager.cpp as I posted image above.

Callstack:

![image](https://github.com/pelicanmapping/rocky/assets/6330648/625d7996-5b8b-4f20-9896-30f9e2ae9bc2)

**Reply by gwaldron** [reply]
Strange. Not able to reproduce here with the main branch. I wonder if building in RelWithDebInfo config would make a difference. I have never personally tried to build VSG or Rocky in Debug config.

**Comment by AnyOldName3** [comment]
If the problem's what I think it is, then it's fixed by the vendor-neutral barycentric coordinate commit that's made it into the VSG 1.1.2 PR. Rocky's set up to use an Nvidia-specific GLSL extension, and requests it when the Vulkan context is created, so the VSG errors out if there's no Nvidia GPU present. There's an identical vendor-neutral extension that also works on many GPUs from other vendors, and all the same Nvidia GPUs provided they're using a driver from 2022 or later.

**Reply by gwaldron** [reply]
OK, I was not aware of this vendor-neutral extension. Thanks for pointing that out.
We are getting ready to update to VSG 1.1.4 and hopefully can resolve this issue at the same time.

**Reply by ComradeMashkov** [reply]
It's a real problem. Rocky does work for me in my PC with NVIDIA GPU, but it does NOT work in my laptop without NVIDIA GPU.

**Reply by ComradeMashkov** [reply]
I tried [this](https://github.com/pelicanmapping/rocky/pull/28/commits/9777615984cf5ef7f6b53a76bfa9851096350a69) commit and it didn't help :(
Can you point it out what extension should I use for laptops with just Intel Iris Xe Graphics (Intel Core™ i7-1260P Processor)?

**Reply by AnyOldName3** [reply]
I just checked, and that chip doesn't support barycentric coordinates at all, at least with drivers that are currently available for Windows or Linux. Maybe you could ask Intel to add support, but if it's a hardware limitation rather than it just not being exposed, then that's not possible. I guess Rocky will need to fall back to drawing wireframes via a second pass with a different `VkPolygonMode` when the extension isn't present instead of dying at launch.

**Reply by ComradeMashkov** [reply]
Well that makes sense. I also didn't find any information about barycentric coordinates support for this CPU. However, I'm lack of information and knowledge in Vulkan and VSG, so I can't realise it by myself even without checking whether extension presented or not. That's why I'll be very grateful if anyone can solve this problem for machines with such a problem.

**Reply by gwaldron** [reply]
Perhaps the best move for now it to simply ifdef out the barycentric stuff, maybe with a cmake flag. It is not mission-critical anyway. Thoughts?

**Reply by ComradeMashkov** [reply]
I will give it a try and publish my results in this comment section.

**Reply by ComradeMashkov** [reply]
Btw, @gwaldron, there are a lot of barycentric stuff in the source code (in `.frag` files, `weemesh.h`, `DisplayManager.cpp`). Which one should I ifdef?

**Reply by ComradeMashkov** [reply]
Well I did it! I commented bary stuff in `.frag` files and in `DisplayManager.cpp`. Thanks y'all!

**Reply by AnyOldName3** [reply]
> I also didn't find any information about barycentric coordinates support for this CPU. However, I'm lack of information and knowledge in Vulkan and VSG, so I can't realise it by myself even without checking whether extension presented or not.

These days I use this site to see which things support which extensions. Here's a link specifically for Iris Xe on Windows with the 101.5534 driver: https://vulkan.gpuinfo.org/displayreport.php?id=30782#extensions

**Reply by gwaldron** [reply]
Which was causing the crash? The use of `GL_NV_fragment_shader_barycentric` in the shader without an `ifdef`, or the code in `DisplayManager.cpp`?

**Reply by AnyOldName3** [reply]
If you've not got the extension, then as `DisplayManager.cpp` tells the VSG that it's a hard requirement, a Vulkan context won't be created, and when `DisplayManager` tries using the context, it won't work. It doesn't get far enough to try using the shaders.

**Reply by gwaldron** [reply, marked as answer]
Thanks. I pushed a couple changes that will hopefully (a) check whether the extension is supported before requesting it, and (b) check in all shaders whether it's been enabled.

**Reply by kahlenberg** [reply]
Thank you all,
It now works on my machine :)
I can build and run the applications.
